src/types/ed25519.rs

```diff
@@ -57,48 +59,130 @@ impl Ed25519Keypair {
+        self.public_key
     }
 
     pub fn public_key_encoded(&self) -> &str {
         &self.encoded_public_key
     }
 
     pub fn sign(&self, message: &[u8]) -> Ed25519Signature {
-        self.secret_key.sign(message, self.public_key())
+        self.secret_key.sign(message, &self.public_key())
+    }
+}
+
+/// An Ed25519 secret key, used to create digital signatures.
+#[derive(Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct Ed25519SecretKey(SecretKey);
+
+impl Ed25519SecretKey {
+    /// Create a new random `Ed25519SecretKey`.
+    pub fn new() -> Self {
+        let mut rng = thread_rng();
+        let key = SecretKey::generate(&mut rng);
+
+        Self(key)
+    }
+
+    /// Get the byte representation of the secret key.
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        self.0.as_bytes()
+    }
+
+    /// Try to create a `Ed25519SecretKey` from a slice of bytes.
+    pub fn from_slice(bytes: &[u8]) -> Result<Self, crate::KeyError> {
+        let key = SecretKey::from_bytes(bytes).map_err(SignatureError::from)?;
+
+        Ok(Self(key))
+    }
+
+    /// Convert the secret key to a base64 encoded string.
+    ///
+    /// This can be useful if the secret key needs to be sent over the network
+    /// or persisted.
+    ///
+    /// **Warning**: The string should be zeroized after it has been used,
+    /// otherwise an unintentional copy of the key might exist in memory.
+    pub fn to_base64(&self) -> String {
+        base64_encode(self.as_bytes())
+    }
+
+    /// Try to create a `Ed25519SecretKey` from a base64 encoded string.
+    pub fn from_base64(key: &str) -> Result<Self, crate::KeyError> {
+        let mut bytes = base64_decode(key)?;
+        let key = Self::from_slice(&bytes);
+
+        bytes.zeroize();
+
+        key
+    }
+
+    /// Get the public key that matches this `Ed25519SecretKey`.
+    pub fn public_key(&self) -> Ed25519PublicKey {
+        Ed25519PublicKey(PublicKey::from(&self.0))
+    }
+
+    /// Sign the given slice of bytes with this `Ed25519SecretKey`.
+    ///
+    /// The signature can be verified using the public key.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use vodozemac::{Ed25519SecretKey, Ed25519PublicKey};
+    ///
+    /// let secret = Ed25519SecretKey::new();
+    /// let message = "It's dangerous to go alone";
+    ///
+    /// let signature = secret.sign(message.as_bytes());
+    ///
+    /// let public_key = secret.public_key();
+    ///
+    /// public_key.verify(message.as_bytes(), &signature).expect("The signature has to be valid");
+    /// ```
+    pub fn sign(&self, message: &[u8]) -> Ed25519Signature {
+        let expanded = ExpandedSecretKey::from(&self.0);
+        Ed25519Signature(expanded.sign(message, &self.public_key().0))
+    }
+}
+
+impl Default for Ed25519SecretKey {
+    fn default() -> Self {
+        Self::new()
     }
 }
 
 #[derive(Serialize, Deserialize)]
-enum Ed25519SecretKey {
+enum SecretKeys {
```

Comment:
Why the plural here?

Reply:
It avoids a name clash with `ed25519_dalek::SecretKey`.

Feel free to change it if you feel that strongly about it.